### PR TITLE
Fix for < > (non HTML) tags disappearing in WYSIWYG

### LINKED ISSUE #12931: WYSIWYG and swithcing to markdown causing issues with deletion of text that have "<" and ">"
- Azure Data Studio Version:
Version: 1.23.0 (system setup)
Commit: d296b6397e0acfddc57e9085e736e084969cdaeb
Date: 2020-10-13T23:03:09.799Z
VS Code: 1.49.0
Electron: 9.2.1
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Windows_NT x64 10.0.19042

Steps to Reproduce:

1. Create a text cell in a notebook.
2. Type `\<testing1\>` in the **Split View** mode.
3. Switch to **Rich Text** mode. Exit out of editing mode.
4. Add a new text underneath the line - say "awesome". Switch to Markdown. Exit out of editing mode.
What you'll see:
The rich text mode removed the `\<` and`\>`.
Going back and forth will cause removal of the entire `\<testing1\>` word. 

This is a breaking change for existing users that may be using this form a lot.
 
BreakingChangeWithLessThanGreaterThan

Expected: no removal of `\<` and`\>` in markdown when switching back and forth. No removal of texts.

### PR DESCRIPTION
**Pull request by VasuBhog**
This PR fixes #12931. I plan on fixing the failing tests, just want to get feedback on PR thus far. 

Tags_gif

### CODE REVIEW

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts [hunk 1]
```diff
@@ -150,3 +219,20 @@ export function findPathRelativeToContent(notebookFolder: string, contentPath: U
 	}
 	return '';
 }
+
+export function escapeAngleBrackets(textContent: any) {
```
Comment (kburtram) [inline src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:223]: why is this any type instead of string?

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts [hunk 2]
```diff
@@ -150,3 +219,20 @@ export function findPathRelativeToContent(notebookFolder: string, contentPath: U
...
+	let text = textContent;
```
Comment (kburtram) [inline src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:224]: use a type when defining variables?

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts [hunk 3]
```diff
@@ -131,6 +134,72 @@ export class HTMLMarkdownConverter {
...
+					}
+				});
+				return '\n\n' + node.innerHTML.replace(/&lt;/gi, '<').replace(/&gt;/gi, '>').replace(/&nbsp;/gi, '') + '\n\n';
+			}
+		});
+		this.turndownService.addRule('heading', {
+			filter: ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'],
+			replacement: function (content, node, options) {
+				let hLevel = Number(node.nodeName.charAt(1));
+				let excapedText = escapeAngleBrackets(content);
```
Comment (kburtram) [inline src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:154]: typo here

File: src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts [hunk 4]
```diff
@@ -152,4 +152,32 @@ suite('HTML Markdown Converter', function (): void {
 		htmlString = '<ol><li>Test<ol><li>Test2</li></ol><li>Test3</li></ol>';
 		assert.equal(htmlMarkdownConverter.convert(htmlString), `1. Test\n    1. Test2\n2. Test3`, 'Basic ordered item test failed');
 	});
+	test('Should keep < > tag', () => {
+		htmlString = '&lt;test&gt';
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '<test>', 'Non-HTML tag test failed to escape');
+		htmlString = '&lt;test&gt<span style="background:red">test</span>&lt;test&gt';
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '<test><span style="background:red">tes</span><test>', 'Non-HTML tag inside span tag test failed to escape');
```
[the remark this answers is not in the document]
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:159]: thanks for catching this, ill make it more clear as well.

**Review by lucyzhang929 — APPROVED** [review]

**Review by barbaravaldez — APPROVED** [review]
Tested and it works well 😄. Just make sure to create an issue for supporting anchor tags in tables and a feature request for escaping anchor tags when user starts in WYSIWYG.

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts [hunk 5]
```diff
@@ -150,3 +219,20 @@ export function findPathRelativeToContent(notebookFolder: string, contentPath: U
 	}
 	return '';
 }
+
+export function escapeAngleBrackets(textContent: string) {
+	let text: string = textContent;
+	if (text.includes('<u>') || text.includes('<mark>') || text.includes('style')) {
```
Comment (corivera) [inline src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:225]: Are there any issues here if the tags contain "style" in their enclosed text, rather than as just an attribute?

Comment (corivera) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:225]: Also, what about a really long html string that contains <mark> or <u> tags in some places, but has text without those tags in others? Would that still be getting escaped properly? Like if you have underlined text before a list, for example.

Comment (corivera) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:225]: Discussed offline. The turndown replacement function deals with individual elements, so we don't need to worry about weird layering like this.

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:225]: Yup, all those scenarios should escape angle brackets properly since turndown breaks up HTML elements and the rules that are I introduced to ensure each Markdown Toolbar actions escapes the non-HTML tag, regardless if they contain multiple HTML tags.

Comment (corivera) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:225]: Although the "style" text comment is still valid.

**Review by corivera — CHANGES_REQUESTED** [review]
We might be hitting an edge case with style attributes, since we only look for the word "style" in the whole html string.

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts (hunk 5 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:225]: The edge case seems to be only when someone would want escape \<style\> properly as text. I'll make sure to update the condition specifically for "style".

**Comment by VasuBhog** [comment]
> We might be hitting an edge case with style attributes, since we only look for the word "style" in the whole html string.

@corivera Fixed and added edge case unit test.

File: src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts [hunk 6]
```diff
@@ -153,6 +154,37 @@ suite('HTML Markdown Converter', function (): void {
...
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '<u>&lt;Underline_test&gt;</u>', 'Basic underline non-HTML tag test failed to escape');
+		htmlString = '<ul><li>&lt;test&gt;</li></ul>';
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '- \\<test\\>', 'Basic unordered list non-HTML tag item test failed to escape');
+		htmlString = '<ol><li>&lt;test&gt;</li></ol>';
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '1. \\<test\\>', 'Basic ordered list non-HTML tag item test failed to escape');
+		htmlString = '<mark>&lt;test&gt;</mark>';
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '<mark>&lt;test&gt;</mark>', 'Basic highlighting Non-HTML tag test failed to escape');
+		htmlString = '<mark><h1>&lt;test&gt;</h1></mark>';
+		assert.equal(htmlMarkdownConverter.convert(htmlString), '<mark><h1>&lt;test&gt;</h1></mark>', 'Non-HTML tag inside multiple html tags test failed to escape');
+		htmlString = '<p>&lt;style&gt</p>';
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:184]: What about `<p>style</p>`?

Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:184]: That would just return the text 'style' as expected.

Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:184]: I'm saying that the content contains the string `<p>style</p>` - which if I'm understanding the logic correctly would not escape the <'s and thus still have the bug.

**Review by corivera — APPROVED** [review]

File: src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts (hunk 6 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:184]: So below is how it WYSIWYG and Markdown should treat 'style', and '<style>'. The above example you have, `<p>style</p>` would be the just the text. You would have to include either the escape `\\` or `&lt;` and `&gt;` to keep the <'s. Therefore, there should not be a bug.

image

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts [hunk 7]
```diff
@@ -150,3 +219,20 @@ export function findPathRelativeToContent(notebookFolder: string, contentPath: U
 	}
 	return '';
 }
+
+export function escapeAngleBrackets(textContent: string, node?) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:223]: Node should have a type

Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:223]: And the function should also specify void as the return type

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts [hunk 8]
```diff
@@ -150,3 +219,20 @@ export function findPathRelativeToContent(notebookFolder: string, contentPath: U
...
+	}
+	let mapTags = { '<': '\\<', '>': '\\>' };
+
+	let escapedText = text.replace(/<|>/gi, function (matched) {
+		return mapTags[matched];
+	});
+	return escapedText;
+}
+
+function repeat(character, count) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:236]: Return type for the function

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts (hunk 7 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts:223]: I was able to refactor and found I don't utilize node anymore. Will delete.

File: src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts (hunk 6 above)
Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:184]: Ok, so what about `\<styletag\>`? Is that also still escaped correctly? 

I'm not even sure I understand why u, mark and style are special in this regard - why are you having to skip escaping the brackets for only those ones?

File: src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts (hunk 8 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts]: refactored the repeat function, to use string.repeat instead.

File: src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts (hunk 6 above)
Comment (VasuBhog) [reply src/sql/workbench/contrib/notebook/test/browser/htmlMarkdownConverter.test.ts:184]: So the specific edge case we were tackling was style, and it seems that this would be an even further edge case that should be fixed outside this PR. 

For your other question, when looking over the WYSIWYG code you can notice that there is a keep setting that we choose to include which allows us to keep certain tags. https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts#L24
You can read about it more here: https://github.com/microsoft/azuredatastudio/blob/main/src/sql/workbench/contrib/notebook/browser/htmlMarkdownConverter.ts#L24

**Review by Charles-Gagnon — APPROVED** [review]
